### PR TITLE
#1904 mobile sign in page

### LINKED ISSUE #1904: add ability to sign in on mobile
we currently don't have a way for users to log in on mobile, which prevents us from knowing which validations are coming from researchers.

### PR DESCRIPTION
**Pull request by rpechuk**
Added two new views that are mobile friendly for sign in and sign up. Resolves #1904.

Screenshots:
image
image

### CODE REVIEW

**Comment by misaugstad** [comment]
@rpechuk could you add the "Resolves #..." and some screenshots to PR description?

**Comment by misaugstad** [comment]
A couple issues:
1. The sign in button remains even if you are already signed in. Should probably be replaced by a username/sign out button.
1. The formatting of the timestamp in the footer is off for me. It looks like "2020.07.03 05:03:27.627522" instead of "July 2, 2020"
IMG_7400

1. Both sign in and sign up zoom in slightly when you click on a text box, which is kind of annoying bc then you need to zoom out after to see the full page again. Maybe take a look at the thread in @shokiami 's PR #2171 where he dealt with this for the feedback box on mobile.
IMG_7404

1. For the sign up page, the bottom kind of gets cut off on my small iPhone SE. The behavior is a little weird though. It starts scrolled up too high. Then when I start scrolling down, there's an extra white bit at the top that covers the page. But then when I scroll a 2nd time it goes away and I can get it to look the way it should. I really just seems like there is some white space at the top when the page loads that shouldn't be there, but then it goes away? Note that this isn't an important thing to fix. It's worth a little bit of time, but not much considering most people don't have tiny phones like me these days.
IMG_7401
---
IMG_7402
---
IMG_7403

**Comment by rpechuk** [comment]
Sounds good thank you for the feedback i'll start looking into all those issues now

**Comment by rpechuk** [comment]
For the first issue I've been looking through the code and trying to understand how to see if the user is signed in and get their username and I can't really figure out how to do both of these so I was wondering if you could help point me in the right direction.

For the second issue I'm also not sure what's going on because when i'm on any sort of emulator it works perfectly fine, however when I switch over to my phone it breaks, even though I didn't touch the footer code at all.

The third issue I fixed and I believe that I found maybe a little less hacky solution which should be in place for everything including the page that Sho was working on.

For the last issue, i'm not able to replicate it at all but I think that what is causing this "issue" is the fact that I added padding to the top of the page so that the content wasn't pressed up against the top (as that doesn't look good)

**Comment by misaugstad** [comment]
>For the first issue I've been looking through the code and trying to understand how to see if the user is signed in and get their username and I can't really figure out how to do both of these so I was wondering if you could help point me in the right direction.

You can see an example of this being done in navbar.scala.html to show the username in the navbar. You can see at the top that a `user: Option[User]` is being passed in. Then later it checks if it is not an anonymous user with `@if(user && user.get.role.getOrElse("") != "Anonymous")`, and if it's a non-anonymous user, it gets their username with `user.get.username`.

>For the second issue I'm also not sure what's going on because when i'm on any sort of emulator it works perfectly fine, however when I switch over to my phone it breaks, even though I didn't touch the footer code at all.

To help narrow this down, the code that formats the timestamp there is in TimestampLocalization.js, which is included and run at the end of main.scala.html. Not sure why it breaks specifically on the phone off the top of my head though.

This reminds me though: do the pages work in Spanish?

>The third issue I fixed and I believe that I found maybe a little less hacky solution which should be in place for everything including the page that Sho was working on.

Nice!

>For the last issue, i'm not able to replicate it at all but I think that what is causing this "issue" is the fact that I added padding to the top of the page so that the content wasn't pressed up against the top (as that doesn't look good)

Does my last image look correct? Or is that what it looks like without padding?

**Comment by misaugstad** [comment]
you'll want to be careful resolving conflicts on this PR. since your navbar PR accidentally took code from this one, the default merge will probably undo those changes here and you'll want to look through and make sure everything is still there.

**Comment by misaugstad** [comment]
This is still happening on my phone:
>Both sign in and sign up zoom in slightly when you click on a text box, which is kind of annoying bc then you need to zoom out after to see the full page again. Maybe take a look at the thread in @shokiami 's PR #2171 where he dealt with this for the feedback box on mobile.

This is also still happening on my phone:
>The formatting of the timestamp in the footer is off for me. It looks like "2020.07.03 05:03:27.627522" instead of "July 2, 2020"

**Comment by rpechuk** [comment]
That's super weird cause neither of those things are happening on my phone, and it's not even just the emulator they just aren't happening.

**Comment by misaugstad** [comment]
For the zooming issue, did you try doing whatever Sho had done in his PR?

**Comment by rpechuk** [comment]
No it's just not doing it for me and it never was so I forgot about it i'll look into it tmr though

File: app/controllers/UserController.scala
Comment (misaugstad) [inline app/controllers/UserController.scala:42]: You can get rid of the `@return` since it really isn't saying anything here

File: app/controllers/UserController.scala
Comment (misaugstad) [inline app/controllers/UserController.scala:66]: Same here

File: app/views/mobileValidate.scala.html
Comment (misaugstad) [inline app/views/mobileValidate.scala.html:65]: Don't think this blank line improves readability at all

File: app/views/navbar.scala.html
Comment (misaugstad) [inline app/views/navbar.scala.html:82]: I don't think any of these added lines do anything either. Not sure if you added them on purpose or if they somehow got added because of a merge

File: app/views/main.scala.html
Comment (misaugstad) [inline app/views/main.scala.html:63]: What does this new piece do?

File: app/views/main.scala.html
Comment (misaugstad) [inline app/views/main.scala.html:69]: I think it's kind of confusing to have this inside that outer if block. I realize that it is technically slightly less efficient to do this check outside of the outer if statement, but I think it reads more clearly that way.

File: app/views/signInMobile.scala.html
Comment (misaugstad) [inline app/views/signInMobile.scala.html:7]: Again, what does this piece do? No need to answer if you answer above.

File: app/views/signInMobile.scala.html
Comment (misaugstad) [inline app/views/signInMobile.scala.html:13]: Can you move all of the style stuff out of here and into CSS? I created a new `mobileAuthenticate.css` file that you can add stuff into if it is most appropriate in there.

File: app/views/signInMobile.scala.html
Comment (misaugstad) [inline app/views/signInMobile.scala.html:52]: missing newline at end of file

File: public/javascripts/SVValidate/mobile/mobileValidate.css
Comment (misaugstad) [inline public/javascripts/SVValidate/mobile/mobileValidate.css:313]: missing newline

**Review by misaugstad — CHANGES_REQUESTED** [review]
Added mostly minor comments. In addition, I noticed this issue in the Spanish language version:
Screenshot from 2020-11-24 16-41-46

**Comment by misaugstad** [comment]
Also I believe that I've fixed the timestamp issue and the zoom in issue. Can you make sure to test and see that I didn't mess up how those look for you?

File: app/views/main.scala.html
Comment (rpechuk) [reply app/views/main.scala.html:63]: it fixes an issue having to do with the page loading in zoomed in

File: app/views/signInMobile.scala.html
Comment (rpechuk) [reply app/views/signInMobile.scala.html:7]: same thing as above, it for some reason only works if it's in both places so I just left it

File: conf/messages.en
Comment (rpechuk) [reply conf/messages.en:195]: not sure what you mean here if you're referring to `authenticate.non.member.mobile` then the readon I have a mobile version it for routing issues.

File: conf/messages.en
Comment (rpechuk) [reply conf/messages.en:194]: Not sure what you mean here as well, the reason it is different is because of routing:
`<a href="/signUpMobile">`

File: app/views/signInMobile.scala.html
Comment (rpechuk) [reply app/views/signInMobile.scala.html:13]: I don't see any `mobileAuthenticate.css` file i'm not sure if that's on my end or not. I don't think it is as I did git pull

Comment (misaugstad) [reply app/views/signInMobile.scala.html:13]: It's in `public/stylesheets/mobileAuthenticate.css`. Also just to make sure, are you using the `1904-Mobile-Sign-In-Page` branch? And not the old `1904-Mobile-Sign-In` or `1904-Sign-In-Mobile` branches? I have since deleted those old branches from Github because I accidentally checked out the wrong one once and got confused. But you might still have it on your local machine. If you were accidentally on the wrong branch, that would also explain the .ivy2 issues you were having.

File: conf/messages.en
Comment (misaugstad) [reply conf/messages.en:195]: No I'm referring to the `authenticate.username` translation. We already have a translation at line 14 `username = Username`.

File: app/views/main.scala.html
Comment (misaugstad) [reply app/views/main.scala.html:63]: Okay, it didn't fix it on my phone so I added a new fix. Could you test out removing this fix and seeing if it still works on your phones?

File: conf/messages.en
Comment (misaugstad) [reply conf/messages.en:194]: Ahh, my bad. I was just thinking about the signing up through the navbar and forgot about the other actual sign in/up pages. So the wording is obviously good, but could you change the translations to take in the link as a parameter? So I think the translation would probably look like:
```
authenticate.non.member.mobile = Not a member? <a href="{0}">Sign up now</a>
```

Then you would pass in the URL in the scala.html. There are examples elsewhere in the code you can go off of.

Then of course do that for both the sign in and sign up translations.

File: app/views/main.scala.html
Comment (rpechuk) [reply app/views/main.scala.html:63]: Removed and tested works fine!

File: conf/messages.en
Comment (rpechuk) [reply conf/messages.en:195]: Fixed in EN and ES

File: conf/messages.en
Comment (rpechuk) [reply conf/messages.en:194]: Fixed this issue. This shows up a lot throughout the code so if you want me to go through and fix every occurrences let me know.

File: app/views/signInMobile.scala.html
Comment (rpechuk) [reply app/views/signInMobile.scala.html:7]: Got rid of this as per other issue

File: app/views/signInMobile.scala.html
Comment (rpechuk) [reply app/views/signInMobile.scala.html:13]: Fixed all occurrences of this issue

File: conf/messages.en
Comment (misaugstad) [reply conf/messages.en]: Oh, yes please! That's kind of what I meant. Since the only difference between the mobile and desktop translations is the link, you could have just a single translation if the link is always passed in as a parameter.

**Comment by misaugstad** [comment]
Okay I fixed one two small things
1. The code did not compile because in one of the instances where you passed in the URL string as a parameter for the translation you put the parameter in the wrong places
2. Changed a CSS class name from `grayButtonMobile` to `gray-button-mobile`

So I fixed those, but we have a new issue. This is what the mobile sign in/up pages are looking like for me now:
Screenshot from 2021-01-25 13-55-46

Were you having this issue too? If not, can you check to see if you are now? I even went back to an old commit before I merged develop in and found the same issue. I think it may have been introduced with commit f79bba4?

**Comment by rpechuk** [comment]
I've been trying my best to re-create this issue and I can't seem to do it. I was never encountering this issue so I have no clue what could be causing it. I'm looking into it right now and i'll keep you posted

**Comment by misaugstad** [comment]
That's wild that it happens to me every time, even in Chrome dev tools, but it isn't happening for you at all. Ugh. I even tried clearing my cache and it's still happening.

**Comment by rpechuk** [comment]
@misaugstad not sure what's cause this issue. The only thing that I can think of that would help at all is the meta tag that we removed. I'll leave it up to you to test however since I can't do it.

**Comment by misaugstad** [comment]
Yep. That did it. So sorry Ron :sweat_smile: I led you on a wild goose chase, my bad :/ It's working much better now.

**Review by misaugstad — APPROVED** [review]
This all looks good now. Thank you for sticking with this one for so long. And sorry again for asking you to remove the code that did help to solve the problem :sweat_smile:

**Pull request merged**
